src/main/java/htsjdk/samtools/cram/build/SliceFactory.java

```diff
@@ -0,0 +1,333 @@
+                        sliceRecordCounter));
+        return sliceRecordCounter + sliceSAMRecords.size();
+    }
+
+    /**
+     * Get the CRAM records accumulated for the current Slice.
+     *
+     * @return the list of all CRAMRecords
+     */
+    public List<CRAMRecord> getCRAMRecords() {
+        // Create a list of ALL reads from all accumulated slices (used to create the container
+        // compression header, which must be presented with ALL reads that will be included in the
+        // container, no matter how they may be distributed across slices). So if more than one slice
+        // entry has been accumulated, we need to temporarily stream all the records into a single
+        // list to present to compressionHeaderFactory.
+        return cramRecordSliceEntries.size() > 1 ?
+                cramRecordSliceEntries.stream().flatMap(e -> e.records.stream()).collect(Collectors.toList()) :
+                cramRecordSliceEntries.get(0).getRecords();
+    }
+
+    public int getNumberOfSliceEntries() {
+        return cramRecordSliceEntries.size();
+    }
+
+    public List<Slice> getSlices(
+            final CompressionHeader compressionHeader,
+            final long containerByteOffset) {
+        final List<Slice> slices = new ArrayList<>(cramRecordSliceEntries.size());
+        for (final SliceStagingEntry sliceStagingEntry : cramRecordSliceEntries) {
+            final Slice slice = new Slice(
+                    sliceStagingEntry.getRecords(),
+                    compressionHeader,
+                    containerByteOffset,
+                    sliceStagingEntry.getGlobalRecordCounter()
+            );
+            slice.setReferenceMD5(cramReferenceRegion.getCurrentReferenceBases());
+            slices.add(slice);
+        }
+        cramRecordSliceEntries.clear();
+        return slices;
+    }
+
+    private final List<CRAMRecord> convertToCRAMRecords(final List<SAMRecord> samRecords, final long sliceRecordCounter) {
+        long recordIndex = sliceRecordCounter;
+        final List<CRAMRecord> cramRecords = new ArrayList<>();
+        for (final SAMRecord samRecord : samRecords) {
+            int referenceIndex = samRecord.getReferenceIndex();
+            final CRAMRecord cramRecord = new CRAMRecord(
+                    CramVersions.DEFAULT_CRAM_VERSION,
+                    encodingStrategy,
+                    samRecord,
+                    cramReferenceRegion.getReferenceBases(referenceIndex),
+                    recordIndex++,
+                    readGroupNameToID);
+            cramRecords.add(cramRecord);
+        }
+        resolveMatesForSlice(cramRecords);
+        return cramRecords;
+    }
+
+    //Update mate pairing
+    private void resolveMatesForSlice(final List<CRAMRecord> cramRecords) {
+        if (!coordinateSorted) {
+            for (final CRAMRecord r : cramRecords) {
+                r.setDetached(true);
```

Comment:
Why is this only in the `! coordinateSorted` case? Shouldn't we always start with the detached assumption?